firedrake/assemble.py

```diff
@@ -644,8 +666,6 @@ class FormAssembler(abc.ABC):
     """
 
     def __init__(self, form, tensor, bcs=(), form_compiler_parameters=None, needs_zeroing=True, weight=1.0):
-        assert tensor is not None
```

Comment:
I don't like that you are breaking this assertion. You are sort of building a dummy `TwoFormAssembler` just to get some properties from the parloops. Could you not get that information some other way? Perhaps by refactoring `ParloopBuilder`?

Reply:
Sparsity construction should basically trace the actual assembly process to be performed (with some abstraction), so I think we should be able to use the same infrastructure, i.e., `FormAssembler`; as we make `ParloopBuilder`s via `FormAssembler` in the actual assembly, we should do the same when making sparsity instead of inventing a different way to make `ParloopBuilder`s only for sparsity. Indeed, I think, `FormAssembler` is nicely reusable in sparsity construction, except possibly for the `tensor is not None` check. I think `FormAssembler` would be cleaner if it did not carry `self._tensor` at all and made `assemble()` method take `tensor` as argument, as `self._tensor` is only used in limited places; I at least included `assert self._tensor is not None` at the top of `assemble()` method.

Reply:
Maybe `FormAssembler` can carry `self._tensor`, but we should probably clarify where it is required, as, apart from `self._tensor`, `FormAssembler` is reusable in sparsity construction.

Reply:
I suppose that it doesn't need to have a `_tensor` attribute. We could perhaps make it an argument of `.assemble()`? I have never liked how I [do this](https://github.com/firedrakeproject/firedrake/blob/master/firedrake/assemble.py#L691).

> Sparsity construction should basically trace the actual assembly process to be performed (with some abstraction), so I think we should be able to use the same infrastructure

This does make sense, but I think the class hierarchy may need tweaking. For instance, `FormAssembler` has an `assemble()` method that does not make sense for a `SparsityConstructor`.

Sorry for being such a stickler here, but assembly is sufficiently complicated that I think we need to be careful about adding any extra "hacks".

Reply:
We could take `_tensor` out, but it turned out that it is currently used to make lgmaps, where we use the underlying `PETSc.Mat`, so I think we should do that in a separate PR.

I added another commit to separate the actual assembly.